pytorch_lightning/utilities/data.py

```diff
@@ -18,28 +18,46 @@
 from torch.utils.data import DataLoader, IterableDataset
 
 from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.warnings import WarningCache
 
 BType = Union[torch.Tensor, str, Mapping[Any, "BType"], Iterable["BType"]]
 
+warning_cache = WarningCache()
+
+
+def _extract_batch_size(batch: BType) -> int:
+    if isinstance(batch, torch.Tensor):
+        yield batch.size(0)
+    elif isinstance(batch, str):
+        yield len(batch)
+    elif isinstance(batch, (Iterable, Mapping)):
+        if isinstance(batch, Mapping):
+            batch = batch.values()
+
+        for sample in batch:
+            yield from _extract_batch_size(sample)
+    else:
+        yield 1
+
 
 def extract_batch_size(batch: BType) -> int:
-    """Recursively unpack a batch to find a torch.Tensor.
+    """Unpack a batch to find a ``torch.Tensor``.
 
     Returns:
         ``len(tensor)`` when found, or ``1`` when it hits an empty or non iterable.
     """
-    if isinstance(batch, torch.Tensor):
-        return batch.size(0)
-    if isinstance(batch, str):
-        return len(batch)
-    if isinstance(batch, dict):
-        sample = next(iter(batch.values()), 1)
-        return extract_batch_size(sample)
-    if isinstance(batch, Iterable):
-        sample = next(iter(batch), 1)
-        return extract_batch_size(sample)
-
-    return 1
+    batch_size = None
```

Comment:
```suggestion
    batch_size = 1
```
or annotation as `Optional[int]`

Reply:
it can never be optional ? since above it will return 1 in the last else condition. The reason I initalized it with None to first batch_size since there is a possibility of getting 1 in the first call itself.

Reply:
that's interesting, I didn't see that but that means mypy can infer that it's never None, i.e., that the for loop always runs?

Reply:
I think yes, it will.

Reply:
On my prs, it never wants to be smart...

Reply:
lucky day maybe 😆